pkg/metadata/service_discovery.go

```diff
@@ -15,21 +15,54 @@
 package metadata
 
 import (
+	"errors"
 	"sync"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/prometheus/common/model"
 
 	"github.com/parca-dev/parca-agent/pkg/discovery"
+	"github.com/parca-dev/parca-agent/pkg/process"
 )
 
+type serviceDiscoveryProvider struct {
+	StatefulProvider
+
+	tree *process.Tree
+}
+
+func (p *serviceDiscoveryProvider) Labels(pid int) (model.LabelSet, error) {
+	if p.state == nil {
+		return nil, errors.New("state not initialized")
+	}
+
+	pids, err := p.tree.FindAllAncestorProcessIDsInSameCgroup(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mtx.RLock()
```

Comment:
This function writes state, so this should be a `p.mtx.Lock()`, no?

Reply:
In only reads the state. Do I miss something?

Reply:
You're totally right. It was the stateful provider that was hiding the writing which confused me.